feat: add shop-screen

ahv15 opened the pull request:
These changes allows the items to be bought in multiplayer using the shop screen.

skaldarnar commented on src/main/java/org/terasology/economy/components/PurchasableComponent.java:
```diff
@@ -0,0 +1,18 @@
+     * The cost of buying the item.
+     * If left blank on the prefab, the value component will be used (if it exists).
+     */
+    public int cost = -1;
```
The default value here is -1, but what does that mean? Does the player get money back when they buy it? And why is -1 a better default value than 0?

skaldarnar replied:
And why do we need this at all? This is just a shortcut to avoid having both `Purchasable` and `Value` on an entity, right?

If possible I'd rather have one "marker component" without any fields instead of tracking the same thing in two places.

skaldarnar commented on src/main/java/org/terasology/economy/events/GiveItemTypeEvent.java:
```diff
@@ -0,0 +1,26 @@
+ * Gives the entity to the target entity
+ */
+@ServerEvent
+public class GiveItemTypeEvent implements Event {
```
Given that the name is so similar to `GiveItemEvent` I think we should explain the difference between the two here, and also why we need this event in addition.

skaldarnar commented on src/main/java/org/terasology/economy/events/GiveItemTypeEvent.java:
```diff
@@ -0,0 +1,26 @@
+public class GiveItemTypeEvent implements Event {
+    Prefab targetPrefab;
+
+    public GiveItemTypeEvent() {
```
It might be better to make this constructor package private as it is only meant for internal usage for deserialization, but this event does not make much sense with a `null` prefab.

skaldarnar commented on src/main/java/org/terasology/economy/systems/ItemAuthoritySystem.java:
```diff
@@ -0,0 +1,35 @@
+
+    @ReceiveEvent
+    public void onGiveItemToPlayer(GiveItemTypeEvent event, EntityRef entity) {
+        if (event.getTargetPrefab() != null) {
```
That's a safe guard as I would have put it here, but given the r[ecent discussion](https://discord.com/channels/270264625419911192/702893102481932308/868881134807826502) on `World::setBlock` I'm wondering what @keturn would do here?

As this event with a `null` prefab is most likely a bug we should at least log a WARN message if this case occurs.
If we want to be sure that we notice that the case occurred we should probably even throw an exception here. I'm not sure whether we are able to figure out the original sender of the event with an exception here, but I hope that the stacktrace would reveal that information.

skaldarnar commented on src/main/java/org/terasology/economy/systems/ItemAuthoritySystem.java:
```diff
@@ -0,0 +1,35 @@
+    public void onGiveItemToPlayer(GiveItemTypeEvent event, EntityRef entity) {
+        if (event.getTargetPrefab() != null) {
+            EntityRef item = entityManager.create(event.getTargetPrefab());
+            if (walletAuthoritySystem.isValidTransaction(entity, -item.getComponent(PurchasableComponent.class).cost)) {
```
Can't we get the component from the prefab even before creating the final entity? If the player cannot purchase an item we should probably not create it.

skaldarnar commented on src/main/java/org/terasology/economy/ui/UIInteractionWrapper.java:
```diff
@@ -0,0 +1,100 @@
+
+/**
+ * Wrapper widget used in the shop screen.
+ * Displays a separate widget whilst capturing user interaction with it.
```
The main use case for this is for tooltips, it seems. Can you explain why this is necessary in more detail?

syntaxi replied:
I looked over the prior commits I made and, although I don't recall all the details, it is likely to do with the way user interaction is handled.

We can't just use the `ItemIcon` widget that this contains, because we cannot assign a listener to it
We need to assign a listener to it, because clicking on this widget indicated that this item is selected.
This is visible from the line in ShopScreen#addBlocks
```java
wrapper.setListener(widget -> handleBlockSelected(block));
```
and the line in ShopScreen#addItems
```java
wrapper.setListener(widget -> handlePrefabSelected(item));
```

The listener indicates which shop entry was selected, as well as calling the handler for the specific type, item vs block.

If this is no longer how the code works, and selection is not done by clicking an item in the ship screen, the  this is likely unneeded. However, as far as I am aware, this is not the case and as such, the wrapper is needed.

skaldarnar replied:
Thanks for the explanation!

skaldarnar commented on src/main/java/org/terasology/economy/ui/ShopScreen.java:
```diff
@@ -0,0 +1,227 @@
+            ItemComponent itemComponent = prefab.getComponent(ItemComponent.class);
+            wareDisplay.setIcon(itemComponent.icon);
+        }
+        wareCost.setText("Cost: " + ShopManager.getWareCost(prefab));
```
Please add a `//TODO` comment to make this translatable.

skaldarnar commented on src/main/java/org/terasology/economy/ShopManager.java:
```diff
@@ -0,0 +1,134 @@
+     * @return The price of the ware.
+     */
+    public static int getWareCost(ComponentContainer ware) {
+        int cost = ware.getComponent(PurchasableComponent.class).cost;
```
I'd add a check using Guava preconditions (`checkArgument`)  that `ware` actually is `Purchasable`. Without this check this method will just fail with a NPE.

skaldarnar commented on src/main/java/org/terasology/economy/ShopManager.java:
```diff
@@ -0,0 +1,134 @@
+        blocks.addAll(blockExplorer.getAvailableBlockFamilies());
+        blocks.addAll(blockExplorer.getFreeformBlockFamilies());
+
+        purchasableBlocks = blocks.stream()
+                .map(blockManager::getBlockFamily)
+                .map(BlockFamily::getArchetypeBlock)
+                .filter(block -> block.getPrefab().isPresent())
+                .filter(block -> block.getPrefab().get().hasComponent(PurchasableComponent.class))
+                .collect(Collectors.toSet());
```
Ah, so all blocks with a `Purchasable` component appear in the shop - that actually sounds reasonable :+1: (at least for the start).

syntaxi replied:
That is how I designed it in gooey defense in order to save time. I could easily see alternatives used, especially if multiple shops are to be used in parallel. (eg, metal renagades). In this situation, perhaps a URI for each shop which is referenced inside the purchasable component?

skaldarnar commented on src/main/java/org/terasology/economy/ShopManager.java:
```diff
@@ -0,0 +1,134 @@
+    public void purchaseBlock(Block block) {
+        EntityRef character = localPlayer.getCharacterEntity();
+        EntityRef blockItem = blockItemFactory.newInstance(block.getBlockFamily());
+        if (character.getComponent(CurrencyStorageComponent.class).amount - getWareCost(blockItem) >= 0) {
+            character.send(new WalletTransactionEvent(-getWareCost(blockItem)));
+            inventoryManager.giveItem(character, character, blockItem);
+        }
```
Why do we do the transaction manually here instead of using the new event?

ahv15 committed:
revert changes

jdrueckert commented:
Do I see it correctly, that this now no longer has a dependency on ST?

ahv15 commented:
> Do I see it correctly, that this now no longer has a dependency on ST?

Yup

jdrueckert commented on src/main/java/org/terasology/economy/ShopManager.java:
```diff
@@ -0,0 +1,108 @@
+     * @param ware The ware to get the price for
+     * @return The price of the ware.
+     */
+    public static int getWareCost(ComponentContainer ware) {
+        Preconditions.checkNotNull(ware.getComponent(ValueComponent.class), "Component Container does not contain a Value Component");
+        return ware.getComponent(ValueComponent.class).value;
+    }
```
Why can we drop the negative costs check that was present here in the GD implementation?

ahv15 replied:
before if the PurchasableComponent was -1 that meant that the price was stored in the ValueComponent but over here instead we are solely relying on the ValueComponent for the price.

jdrueckert replied:
Makes me wonder why we differentiate between the two in the first place and why GD even allows for the price to be in either.

skaldarnar replied:
I'd assume this has "historical reasons" and our usual reluctance to use and/or modify existing concepts :thinking:

jdrueckert commented on src/main/java/org/terasology/economy/systems/ItemAuthoritySystem.java:
```diff
@@ -0,0 +1,61 @@
+    public void onGiveItemToPlayer(GiveItemTypeEvent event, EntityRef entity) {
+        if (event.getTargetPrefab() != null && event.getTargetPrefab().hasComponent(ValueComponent.class)) {
+            EntityRef item = entityManager.create(event.getTargetPrefab());
+            if (walletAuthoritySystem.isValidTransaction(entity, -item.getComponent(ValueComponent.class).value)) {
+                entity.send(new WalletTransactionEvent(-item.getComponent(ValueComponent.class).value));
```
Why does the `ItemAuthoritySystem` check the transaction and send the `WalletTransactionEvent` and not the `ShopManager`?

ahv15 replied:
Because the ShopManager is not an authoritative system and these checks are better used in Authoritative Systems.

jdrueckert replied:
Is there a reason, why the ShopManager should stay a non-authoritative system?

ahv15 replied:
This is so that the ShopScreen which is accessible by all players can access the ShopManager.

jdrueckert commented on src/main/java/org/terasology/economy/events/GiveItemTypeEvent.java:
```diff
@@ -0,0 +1,36 @@
+import org.terasology.engine.entitySystem.prefab.Prefab;
+import org.terasology.engine.network.ServerEvent;
+
+/**
+ * Trigger event that sends the required item (as a string/prefab) to the server.
+ * It then creates the item entity and gives the item on the server.
+ */
```
I guess the difference to `GiveItem` here is that the item is not actually "moved" but only a copy of it? I think it would help to actively state the difference here, including a cross-reference.

syntaxi replied:
My understanding is that this event is a trigger, to create and then give an item to the player.

jdrueckert commented on src/main/java/org/terasology/economy/systems/ItemAuthoritySystem.java:
```diff
@@ -0,0 +1,61 @@
+        } else if (event.getBlockURI() != null) {
+            BlockFamily blockFamily = blockManager.getBlockFamily(event.getBlockURI());
+            EntityRef blockItem = blockItemFactory.newInstance(blockFamily);
+            if (walletAuthoritySystem.isValidTransaction(entity, -blockItem.getComponent(ValueComponent.class).value)) {
+                entity.send(new WalletTransactionEvent(-blockItem.getComponent(ValueComponent.class).value));
+                blockItem.send(new GiveItemEvent(entity));
+            }
```
It seems like maybe we could extract some code shared between the prefab and the block case here.

jdrueckert commented on src/main/java/org/terasology/economy/ui/UIInteractionWrapper.java:
```diff
@@ -0,0 +1,101 @@
+ * Displays a separate widget whilst capturing user interaction with it.
+ * When the icon corresponding to the item is clicked, a tooltip containing the item's name and additional information is displayed.
+ * <p>
+ * All size values etc etc are determined by the content widget.
```
```suggestion
 * All size values etc are determined by the content widget.
```

jdrueckert reviewed (changes requested):
We now have code style annotations in the "Files changed" tab. Please have a look and try to resolve them. Resolving issues in files that your PR doesn't touch is optional (yet appreciated :sweat_smile:)

syntaxi commented on src/main/java/org/terasology/economy/systems/ItemAuthoritySystem.java:
```diff
@@ -0,0 +1,61 @@
+import org.terasology.module.inventory.systems.InventoryAuthoritySystem;
+
+@RegisterSystem(RegisterMode.AUTHORITY)
+public class ItemAuthoritySystem extends BaseComponentSystem {
```
I believe that it may be wise to remove this system altogether

Like it was pointed out, it makes more sense for the shop system to check the value, and furthermore I don't see why any of this code has to be done in a separate system.
If we are firing an event, which is only intended to be used for calling this system, then the two should be combined.

ahv15 replied:
The main reason for this system is because the GiveItemEvent seems to work mainly on Authoritative Systems and this becomes necessary for multiplayer, and after discussing this with @skaldarnar and @jdrueckert they suggested creating a separate system so that even MR can eventually use this. (MR also makes use of a separate system for this in a similar manner if I am not mistaken)

skaldarnar replied:
I tried to integrate that (as authority event handler) into the `ShopManager` in 969e482

syntaxi reviewed (commented):
Minor comments and clarifying a question from skaldanar

ahv15 committed:
make requested changes

ahv15 committed:
fix merge conflicts

ahv15 committed:
update

ahv15 committed:
make-changes

jdrueckert commented on src/main/java/org/terasology/economy/events/GiveItemTypeEvent.java:
```diff
@@ -0,0 +1,36 @@
+import org.terasology.engine.entitySystem.prefab.Prefab;
+import org.terasology.engine.network.ServerEvent;
+
+/**
+ * Trigger event that sends the description of the required item (as a string/prefab) to the server.
+ * It then creates the item entity using the description and gives the item on the server.
+ */
```
Seems like there's some comments that got lost here: I think this documentation still does not clearly distinguish `GiveItemType` from the "normal" `GiveItem`

skaldarnar commented on src/main/java/org/terasology/economy/ShopManager.java:
```diff
@@ -0,0 +1,108 @@
+
+    @Override
+    public void postBegin() {
+        BlockExplorer blockExplorer = new BlockExplorer(assetManager);
```
code style nitpick: move this closer to where it is used, i.e., after the `purchasableItems` statement.

skaldarnar commented on src/main/java/org/terasology/economy/components/PurchasableComponent.java:
```diff
@@ -0,0 +1,16 @@
+ * Indicates that the item can be bought and thus will be available in the shop
+ */
+@AddToBlockBasedItem
+public class PurchasableComponent implements Component<PurchasableComponent> {
```
```suggestion
public class PurchasableComponent extends EmptyComponent<PurchasableComponent> {
```

skaldarnar commented on src/main/java/org/terasology/economy/components/PurchasableComponent.java:
```diff
@@ -0,0 +1,16 @@
+ */
+@AddToBlockBasedItem
+public class PurchasableComponent implements Component<PurchasableComponent> {
+    @Override
+    public void copyFrom(PurchasableComponent other) {
+    }
```
```suggestion
```

skaldarnar commented on src/main/java/org/terasology/economy/handler/CurrencyStorageHandler.java:
```diff
@@ -15,13 +15,13 @@
 
-    private Logger logger = LoggerFactory.getLogger(CurrencyStorageHandler.class);
+    private static final Logger logger = LoggerFactory.getLogger(CurrencyStorageHandler.class);
+    private final String currency = "currency";
```
Why change this this way? It looks like it was intended to be a constant value, so just making it static as well would be a better fix?
```suggestion
    private static final String CURRENCY = "currency";
```

ahv15 replied:
I was trying to resolve some of the suggestions given.

skaldarnar commented on src/main/java/org/terasology/economy/ui/UIInteractionWrapper.java:
```diff
@@ -0,0 +1,101 @@
+ * Displays a separate widget whilst capturing user interaction with it.
+ * When the icon corresponding to the item is clicked, a tooltip containing the item's name and additional information is displayed.
+ * <p>
+ * All size values etc are determined by the content widget.
+ */
```
I'd like to add @syntaxi 's explanation (https://github.com/Terasology/Economy/pull/22/files#r696603417) as in-code comment here:

```suggestion
 * All size values etc are determined by the content widget.
 * <p>
 * NOTE: We can't just use the {@code ItemIcon} widget that this contains, because we cannot assign a listener to it.
 * We need to assign a listener to it, because clicking on this widget indicated that this item is selected.
 * This is visible from the line in ShopScreen#addBlocks
 * <pre>
 * wrapper.setListener(widget -> handleBlockSelected(block));
 * </pre>
 * and the line in ShopScreen#addItems
 * <pre>
 * wrapper.setListener(widget -> handlePrefabSelected(item));
 * </pre>
 * The listener indicates which shop entry was selected, as well as calling the handler for the specific type, item vs block.
 */
```

ahv15 committed:
make changes

ahv15 committed:
make changes

skaldarnar committed:
refactor: integrate ItemAuthoritySystem into ShopManager

I think even if a system is registered ALWAYS, we can have a `netFilter` annotated to specific event handlers, like so:

```
@ReceiveEvent(netFilter = RegisterMode.AUTHORITY)
```

This allows us to have code running only on the authority, e.g., for handling purchase of items, while avoiding a completely different system.

skaldarnar committed:
refactor: rename GiveItemTypeEvent >>> PurchaseItemEvent

jdrueckert committed:
Merge branch 'develop' into shop-screen

jdrueckert committed:
fix: add missing imports
